classes/Task/Upgrade/BackupDb.php

```diff
@@ -271,39 +240,134 @@ public function run(): int
+    }
+
+    protected function warmUp(): int
+    {
+        $this->container->getState()->setProgressPercentage(
+            $this->container->getCompletionCalculator()->getBasePercentageOfTask(self::class)
+        );
+
+        $relative_backup_path = str_replace(_PS_ROOT_DIR_, '', $this->container->getProperty(UpgradeContainer::BACKUP_PATH));
+        $report = '';
+        if (!\ConfigurationTest::test_dir($relative_backup_path, false, $report)) {
+            $this->logger->error($this->translator->trans('Backup directory is not writable (%path%).', ['%path%' => $this->container->getProperty(UpgradeContainer::BACKUP_PATH)]));
+            $this->next = 'error';
             $this->setErrorFlag();
 
             return ExitCode::FAIL;
-        } else {
-            $this->container->getState()
-                ->setBackupLoopLimit(null)
-                ->setBackupLines(null)
-                ->setBackupTable(null);
-            if ($found) {
-                $this->logger->info($this->translator->trans('%s tables have been saved.', [$found]));
+        }
+
+        if (!is_dir($this->container->getProperty(UpgradeContainer::BACKUP_PATH) . DIRECTORY_SEPARATOR . $this->container->getState()->getBackupName())) {
+            mkdir($this->container->getProperty(UpgradeContainer::BACKUP_PATH) . DIRECTORY_SEPARATOR . $this->container->getState()->getBackupName());
+        }
+        $this->container->getState()->setDbStep(0);
+        $listOfTables = $this->container->getDb()->executeS('SHOW TABLES LIKE "' . _DB_PREFIX_ . '%"', true, false);
+
+        $tablesToBackup = array_filter(array_map('current', $listOfTables), function ($table) {
+            // Skip tables which do not start with _DB_PREFIX_
+            if (strlen($table) <= strlen(_DB_PREFIX_) || strncmp($table, _DB_PREFIX_, strlen(_DB_PREFIX_)) != 0) {
+                return false;
             }
-            $this->stepDone = true;
-            // reset dbStep at the end of this step
-            $this->container->getState()->setDbStep(0);
 
-            $this->logger->info($this->translator->trans('Database backup done in filename %s. Now upgrading files...', [$this->container->getState()->getBackupName()]));
-            $this->next = 'upgradeFiles';
+            // Ignore stat tables
+            if (in_array($table, $this->getTablesToIgnore())) {
+                return false;
+            }
 
-            return ExitCode::SUCCESS;
+            return true;
+        });
+
+        if (empty($tablesToBackup)) {
+            throw (new UpgradeException($this->translator->trans('No valid tables were found to back up. Backup of database canceled.')))->setSeverity(UpgradeException::SEVERITY_ERROR);
         }
+
+        $tablesToBackup = new Backlog($listOfTables, count($listOfTables));
+
+        $this->container->getFileConfigurationStorage()->save($tablesToBackup->dump(), UpgradeFileNames::DB_TABLES_TO_BACKUP_LIST);
+
+        return ExitCode::SUCCESS;
```

Comment:
By mocking the database it should be possible to add unit tests on this function from hell... what do you think?